recipes/ncbi-cxx-toolkit-public/26/conanfile.py

```diff
@@ -0,0 +1,1438 @@
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+
+class NcbiCxxToolkit(ConanFile):
+    name = "ncbi-cxx-toolkit-public"
+    license = "CC0-1.0"
+    homepage = "https://ncbi.github.io/cxx-toolkit"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "NCBI C++ Toolkit"
+    topics = ("ncbi", "biotechnology", "bioinformatics", "genbank", "gene",
+              "genome", "genetic", "sequence", "alignment", "blast",
+              "biological", "toolkit", "c++")
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake", "cmake_find_package"
+    short_paths = True
+
+    options = {
+        "shared":     [True, False],
+        "fPIC":       [True, False],
+        "with_projects": "ANY",
+        "with_targets":  "ANY"
+    }
+    default_options = {
+        "shared":     False,
+        "fPIC":       True,
+        "with_projects":  "",
+        "with_targets":   "",
+    }
+    NCBI_to_Conan_requires = {
+        "BerkeleyDB":   "libdb/5.3.28@#355aa134e9ee1bda60fa224f8b54c913",
+        "BZ2":          "bzip2/1.0.8@#493645787acdd3909523a6a2773b1b36",
+        "CASSANDRA":    "cassandra-cpp-driver/2.15.3@#27c46dd9dfbb2a3afef30aa2e914f00a",
+        "GIF":          "giflib/5.2.1@#75a273d91a6c63bc0d8a93799ce19acc",
+        "JPEG":         "libjpeg/9d@#956b5e73632aac77088571f574d766c0",
+        "LMDB":         "lmdb/0.9.29@#3bdb51ff4c3649bd865ee3857f0f2e07",
+        "LZO":          "lzo/2.10@#76a80652793ae478a12f6f9a8eb17a89",
+        "MySQL":        "libmysqlclient/8.0.25@#bbe62517865849c442c032cc7aecfd57",
+        "NGHTTP2":      "libnghttp2/1.45.1@#cef166f78349143079e1c667d823b25a",
+        "PCRE":         "pcre/8.45@#b6bae758d36ca2050bb8527ea408eb75",
+        "PNG":          "libpng/1.6.37@#8b23717a9884a2b06884c599072955f4",
+        "SQLITE3":      "sqlite3/3.36.0@#567836d4f4b7b58a4e834bf98288d609",
+        "TIFF":         "libtiff/4.3.0@#28ac662980826ae8ffe8d6236cdd9735",
+        "XML":          "libxml2/2.9.12@#9817dd585ffc6de1479da6d5bcf01fc0",
+        "XSLT":         "libxslt/1.1.34@#47946f5c7abe03d18179b55be67bbabe",
+        "UV":           "libuv/1.42.0@#e1801aae9570062012d94db20338451b",
+        "Z":            "zlib/1.2.11@#683857dbd5377d65f26795d4023858f9",
+        "OpenSSL":      "openssl/1.1.1l@#98215f1d057178dfd8c867950e16a0fe",
+        "ZSTD":         "zstd/1.5.0@#2797bc9a304b2c45e2239c0d4ad15207"
+    }
```

Comment:
Yes, lots of concerns. During this time it was several times that "latest and greatest" revisions broke the build. Using specific revisions was a recommendation I received from one of Conan people - in one of the now hidden comments above.

Reply:
Also I don't see any reason to not require latest version of openssl, zstd, sqlite3 etc

Reply:
We "require" what we have tested and know it works. We cannot update and retest our recipe twice a week

Reply:
> Yes, lots of concerns. During this time it was several times that "latest and greatest" revisions broke the build. Using specific revisions was a recommendation I received from one of Conan people - in one of the now hidden comments above.

It is recommended for the root project (lock file actually), but must not be hardcoded in CCI recipes.

Reply:
can you please give me a link to documentation? I am not sure I understand you. Still, I repeat, updated recipes often break the compilation. They are unreliable

Reply:
> can you please give me a link to documentation?

Which topic? RREV? I don't think it is explicitly explained. Maybe it should.

/cc @jgsogo @SSE4 @uilianries

Reply:
Hi, @gouriano , I wrote about it some days ago. You can start with this FAQ (https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#how-to-protect-my-project-from-breaking-changes-in-recipes), that will soon redirect you to https://github.com/conan-io/conan-center-index/blob/master/docs/consuming_recipes.md

I'll be happy to help or to improve that documentation if needed. As said above, recipe-revs and lockfiles are highly recommended in the consumer project, but not in the recipes. If you want real examples using them in a project, I can shared with you this repository: https://github.com/jgsogo/render_context, you can also have a look to the CI files to check how I use the lockfiles to prevent upstream changes from breaking my builds. I plan to write about it, but didn't have time yet.

Reply:
It is all great. I have studied these docs long ago. it tells me about, "recipe changes breaking a consumer project" (this is what you call "root project", don't you?). This scenario is different though. Here recipe changes break MY recipe. How are you suggesting to handle that??

Reply:
"using your own artifactory instance" - does it mean stop bothering us with your submission? Do not use JFrog ConanCenter and be happy?
"lockfiles" - are they allowed in ConanCenter submissions? So far I thought the answer was "no"

Reply:
@gouriano , I'm giving you the alternatives that can work for your situation (what I understand is your scenario). 

Of course, I would love to improve some things and to have much more resources, but I really think the people here are doing their best, much more than we could request them even if they are employees (there are also many outside contributors 🤟 ). 

I would love to understand better your pain and try to provide a solution that works for you. I'm pushing Conan inside our development (spoiler: backend and microservices are starting to use C++ and packages from ConanCenter) and we have come across some scenarios and bugs, and we need some workarounds,... also I try to use Conan in my own personal projects, trying to understand some difficulties and trying to provide feedback and also workarounds. The projects I share with you do not use anything that is not free and available for everybody.

Of course, if you use this recipe in a company, I'm sure that your project is much bigger than my toy ones, and probably it is also bigger than the one we are working on in the backend.

You can either consume the recipes via your own Artifactory and store there only the recipe_revisions that work for you, or you can use lockfiles and keep using ConanCenter. Those lockfiles won't be in ConanCenter, **those lockfiles need to be in your projects and libraries** (they can be stored in your repositories), and your CIs need to use them whenever they retrieve packages from ConanCenter. We store all past recipe revisions (and binaries) just because of this, I can tell you those are a huge amount of binaries.

Please, share your scenario, try to be more specific and I will be glad to help you with your setup.

Reply:
Our - I am not a person here, I represent an organization. So, I am not supposed to express my personal opinions - So, our goal is simple: make NCBI C++ Toolkit more easily available to people outside of NCBI. The toolkit has long history of development, "traditional" practices, specific design decisions. All of these have nothing to do with Conan. Mapping the toolkit's practices to thats of Conan is not an easy task. We did what we did. It works, and we believe it works good.

When building the toolkit outside of NCBI, the major problem is building 3rd party packages (dependecies). That is, before they even start trying the toolkit, they have to do a lot of things. Conan is supposed to do this "automatically". Which it does very well. 

Of course, our desire is that if our recipe works fine today, it would work fine in a year, or two. The problem with Conan, according to my personal experience, is that it is not so. A recipe may work fine today and then suddenly break tomorrow. Recipe revisions was a good way to fix that. I thought so.

Reply:
I have remove recipe revisions now. So far, it works.